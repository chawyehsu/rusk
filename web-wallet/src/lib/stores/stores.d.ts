import type { Readable } from "svelte/store";
import type { Wallet } from "@dusk-network/dusk-wallet-js";

type TransactionsStoreContent = { transactions: Transaction[] };

type TransactionsStore = Readable<TransactionsStoreContent>;

type WalletStoreContent = {
	balance: {
		maximum: number;
		value: number;
	};
	currentAddress: string;
	error: Error | null;
	initialized: boolean;
	addresses: string[];
	isSyncing: boolean;
};

type WalletStoreServices = {
	clearLocalData: () => Promise<void>;

	clearLocalDataAndInit: (wallet: Wallet) => Promise<void>;

	getStakeInfo: () => Promise<any> & ReturnType<Wallet["stakeInfo"]>;

	// The return type apparently is not in a promise here
	getTransactionsHistory: () => Promise<ReturnType<Wallet["history"]>>;

	init: (wallet: Wallet) => Promise<void>;

	reset: () => void;

	setCurrentAddress: (address: string) => Promise<void>;

	stake: (amount: number, gasPrice: number, gasLimit: number) => Promise<any> & ReturnType<Wallet["stake"]>;

	sync: () => Promise<void>;

	transfer: (to: string, amount: number, gasPrice: number, gasLimit: number) => Promise<any> & ReturnType<Wallet["transfer"]>;

	unstake: (gasPrice: number, gasLimit: number) => Promise<any> & ReturnType<Wallet["unstake"]>;

	withdrawReward: (gasPrice: number, gasLimit: number) => Promise<any> & ReturnType<Wallet["withdrawReward"]>;
};

type WalletStore = Readable<WalletStoreContent> & WalletStoreServices;
