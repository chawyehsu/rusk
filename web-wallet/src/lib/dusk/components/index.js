export { default as Agreement } from "./Agreement/Agreement.svelte";
export { default as Anchor } from "./Anchor/Anchor.svelte";
export { default as AnchorButton } from "./AnchorButton/AnchorButton.svelte";
export { default as Badge } from "./Badge/Badge.svelte";
export { default as Button } from "./Button/Button.svelte";
export { default as Card } from "./Card/Card.svelte";
export { default as CopyField } from "./CopyField/CopyField.svelte";
export { default as Checkbox } from "./Checkbox/Checkbox.svelte";
export { default as CircularIcon } from "./Icon/CircularIcon.svelte";
export { default as ErrorAlert } from "./ErrorAlert/ErrorAlert.svelte";
export { default as ErrorDetails } from "./ErrorDetails/ErrorDetails.svelte";
export { default as Icon } from "./Icon/Icon.svelte";
export { default as Mnemonic } from "./Mnemonic/Mnemonic.svelte";
export { default as ProgressBar } from "./ProgressBar/ProgressBar.svelte";
export { default as QrCode } from "./QrCode/QrCode.svelte";
export { default as Select } from "./Select/Select.svelte";
export { default as Stepper } from "./Stepper/Stepper.svelte";
export { default as Suspense } from "./Suspense/Suspense.svelte";
export { default as Switch } from "./Switch/Switch.svelte";
export { default as Tabs } from "./Tabs/Tabs.svelte";
export { default as Textbox } from "./Textbox/Textbox.svelte";
export { default as Throbber } from "./Throbber/Throbber.svelte";
export { default as Toast } from "./Toast/Toast.svelte";
export { default as Tooltip } from "./Tooltip/Tooltip.svelte";
export { default as Wizard } from "./Wizard/Wizard.svelte";
export { default as WizardStep } from "./Wizard/WizardStep.svelte";
export { default as Words } from "./Words/Words.svelte";
