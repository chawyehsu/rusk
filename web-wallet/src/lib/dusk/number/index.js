export { default as createNumberFormatter } from "./createNumberFormatter";
