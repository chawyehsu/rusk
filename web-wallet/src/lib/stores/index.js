export { default as gasStore } from "./gasStore";
export { default as operationsStore } from "./operationsStore";
export { default as mnemonicPhraseResetStore } from "./mnemonicPhraseResetStore";
export { default as settingsStore } from "./settingsStore";
export { default as walletStore } from "./walletStore";
