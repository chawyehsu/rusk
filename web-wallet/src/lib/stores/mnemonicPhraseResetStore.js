import { writable } from "svelte/store";

/** @type {import('svelte/store').Writable<string[]>} */
const mnemonicPhraseResetStore = writable([]);

export default mnemonicPhraseResetStore;
