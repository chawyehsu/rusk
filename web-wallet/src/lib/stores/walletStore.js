import { get, writable } from "svelte/store";
import { getKey, uniquesBy } from "lamb";

/**
 * @typedef {import("@dusk-network/dusk-wallet-js").Wallet} Wallet
 */

/**
 * @typedef {import("./stores").WalletStoreServices["getTransactionsHistory"]} GetTransactionsHistory
 */

/** @type {Promise<void> | null} */
let syncPromise = null;

/** @type {Wallet | null} */
let walletInstance = null;

const uniquesById = uniquesBy(getKey("id"));

/** @type {import("./stores").WalletStoreContent} */
const initialState = {
	addresses: [],
	balance: {
		maximum: 0,
		value: 0
	},
	currentAddress: "",
	error: null,
	initialized: false,
	isSyncing: false
};

/* eslint-disable-next-line svelte/require-store-callbacks-use-set-param */
const walletStore = writable(initialState);
const { set, subscribe } = walletStore;

/**
 * Defensive code here as the `amount` and
 * `reward` properties can be `undefined` in the
 * returned stake info object.
 *
 * @param {WalletStakeInfo} stakeInfo
 * @returns {WalletStakeInfo}
 */
const fixStakeInfo = stakeInfo => ({
	...stakeInfo,
	amount: stakeInfo.amount ?? 0,
	reward: stakeInfo.reward ?? 0
});

const getCurrentAddress = () => get(walletStore).currentAddress;

/** @type {(action: (...args: any[]) => Promise<any>) => Promise<void>} */
const syncedAction = action => sync().then(action).finally(sync);

/** @type {() => Promise<void>} */
const clearLocalData = async () => walletInstance?.reset();

/** @type {(wallet: Wallet) => Promise<void>} */
const clearLocalDataAndInit = wallet => wallet.reset().then(() => init(wallet));

/** @type {import("./stores").WalletStoreServices["getStakeInfo"]} */
// @ts-expect-error
const getStakeInfo = async () => sync().then(() => walletInstance.stakeInfo(getCurrentAddress()))
	.then(fixStakeInfo);

/** @type {GetTransactionsHistory} */

// @ts-expect-error
const getTransactionsHistory = async () => sync().then(() => walletInstance.history(getCurrentAddress()))
	.then(uniquesById);

function reset () {
	walletInstance = null;
	set(initialState);
}

async function updateAfterSync () {
	const store = get(walletStore);

	// @ts-expect-error
	const balance = await walletInstance.getBalance(store.currentAddress);

	set({
		...store,
		balance,
		isSyncing: false
	});
}

/** @param {Wallet} wallet */
async function init (wallet) {
	walletInstance = wallet;

	const addresses = walletInstance.getPsks();
	const currentAddress = addresses[0];

	set({
		...initialState,
		addresses,
		currentAddress,
		initialized: true

	});
	sync();
}

/** @type {import("./stores").WalletStoreServices["setCurrentAddress"]} */
async function setCurrentAddress (address) {
	const store = get(walletStore);

	return store.addresses.includes(address)
		? Promise.resolve(set({ ...store, currentAddress: address })).then(sync)
		: Promise.reject(new Error("The received address is not in the list"));
}

/** @type {import("./stores").WalletStoreServices["stake"]} */
// @ts-expect-error
const stake = async (amount, gasPrice, gasLimit) => syncedAction(() => {
	if (walletInstance) {
		walletInstance.gasLimit = gasLimit;
		walletInstance.gasPrice = gasPrice;

		return walletInstance.stake(getCurrentAddress(), amount);
	}
});

/** @type {import("./stores").WalletStoreServices["sync"]} */
function sync () {
	if (!walletInstance) {
		throw new Error("No wallet instance to sync");
	}

	if (!syncPromise) {
		const store = get(walletStore);

		set({ ...store, error: null, isSyncing: true });

		syncPromise = walletInstance.sync().then(
			updateAfterSync,
			error => { set({ ...store, error, isSyncing: false }); }
		).finally(() => { syncPromise = null; });
	}

	return syncPromise;
}

/** @type {import("./stores").WalletStoreServices["transfer"]} */
// @ts-expect-error
const transfer = async (to, amount, gasPrice, gasLimit) => syncedAction(() => {
	if (walletInstance) {
		walletInstance.gasLimit = gasLimit;
		walletInstance.gasPrice = gasPrice;

		return walletInstance.transfer(
			getCurrentAddress(),
			to,
			amount
		);
	}

	return null;
});

/** @type {import("./stores").WalletStoreServices["unstake"]} */
// @ts-expect-error
const unstake = async (gasPrice, gasLimit) => syncedAction(() => {
	if (walletInstance) {
		walletInstance.gasLimit = gasLimit;
		walletInstance.gasPrice = gasPrice;

		return walletInstance.unstake(getCurrentAddress());
	}

	return null;
});

/** @type {import("./stores").WalletStoreServices["withdrawReward"]} */
// @ts-expect-error
const withdrawReward = async (gasPrice, gasLimit) => syncedAction(() => {
	if (walletInstance) {
		walletInstance.gasLimit = gasLimit;
		walletInstance.gasPrice = gasPrice;

		return walletInstance.withdrawReward(getCurrentAddress());
	}

	return null;
});

/** @type {import("./stores").WalletStore} */
export default {
	clearLocalData,
	clearLocalDataAndInit,
	getStakeInfo,
	getTransactionsHistory,
	init,
	reset,
	setCurrentAddress,
	stake,
	subscribe,
	sync,
	transfer,
	unstake,
	withdrawReward
};
