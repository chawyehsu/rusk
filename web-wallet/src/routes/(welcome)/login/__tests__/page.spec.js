import {
  afterAll,
  afterEach,
  beforeAll,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";
import { generateMnemonic } from "bip39";
import { get } from "svelte/store";
import { Wallet } from "@dusk-network/dusk-wallet-js";
import { setKey } from "lamb";

import { addresses } from "$lib/mock-data";
import { getAsHTMLElement } from "$lib/dusk/test-helpers";
import * as navigation from "$lib/navigation";
import { settingsStore, walletStore } from "$lib/stores";
import { encryptMnemonic, getSeedFromMnemonic } from "$lib/wallet";
import loginInfoStorage from "$lib/services/loginInfoStorage";
import * as walletService from "$lib/services/wallet";

import Login from "../+page.svelte";

/** @param {HTMLElement} container */
function getTextInput(container) {
  // eslint-disable-next-line no-extra-parens
  return /** @type {HTMLInputElement} */ (
    container.querySelector("[type='password']")
  );
}

describe("Login", async () => {
  const walletGetPsksSpy = vi
    .spyOn(Wallet.prototype, "getPsks")
    .mockResolvedValue(addresses);
  const mnemonic = generateMnemonic();
  const pwd = "some pwd";
  const loginInfo = await encryptMnemonic(mnemonic, pwd);
  const seed = getSeedFromMnemonic(mnemonic);
  const userId = (await new Wallet(seed).getPsks())[0];
  const getErrorElement = () => document.querySelector(".login__error");
  const getWalletSpy = vi.spyOn(walletService, "getWallet");
  const gotoSpy = vi.spyOn(navigation, "goto");
  const initSpy = vi.spyOn(walletStore, "init");

  afterEach(async () => {
    cleanup();
    getWalletSpy.mockClear();
    gotoSpy.mockClear();
    initSpy.mockClear();
    settingsStore.reset();
    walletGetPsksSpy.mockClear();
    walletStore.reset();
  });

  afterAll(() => {
    getWalletSpy.mockRestore();
    gotoSpy.mockRestore();
    initSpy.mockRestore();
    walletGetPsksSpy.mockRestore();
  });

  describe("Mnemonic phrase workflow", () => {
    it("should render the login page with a mnemonic phrase field", () => {
      const { container } = render(Login, {});

      expect(container.firstChild).toMatchSnapshot();
    });

    it("should show an error message if the user enters an invalid mnemonic", async () => {
      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      expect(getErrorElement()).toBeNull();

      await fireEvent.input(textInput, { target: { value: "foo bar" } });
      await fireEvent.submit(form, { currentTarget: form });

      const errorElement = await vi.waitUntil(getErrorElement);
      const selectedText = textInput.value.substring(
        Number(textInput.selectionStart),
        Number(textInput.selectionEnd)
      );

      expect(initSpy).not.toHaveBeenCalled();
      expect(errorElement?.textContent).toMatch(/mnemonic/i);
      expect(textInput).toHaveFocus();
      expect(selectedText).toBe(textInput.value);
    });

    it("should trigger the Reset flow if the user inputs a valid mnemonic with no prior wallet created", async () => {
      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: mnemonic } });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getWalletSpy).toHaveBeenCalledTimes(1);
      expect(getWalletSpy).toHaveBeenCalledWith(seed);
      expect(get(settingsStore).userId).toBe("");
      expect(initSpy).not.toHaveBeenCalled();
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/setup/restore");
    });

    it("should redirect to the dashboard if the entered mnemonic is the last one used", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: mnemonic } });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getWalletSpy).toHaveBeenCalledTimes(1);
      expect(getWalletSpy).toHaveBeenCalledWith(seed);
      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(Wallet));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });

    it("should trim and lower case the entered mnemonic before validating it", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, {
        target: { value: `  \t${mnemonic.toUpperCase()} \t  ` },
      });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getWalletSpy).toHaveBeenCalledTimes(1);
      expect(getWalletSpy).toHaveBeenCalledWith(seed);
      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(Wallet));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });
  });

  describe("Password workflow", () => {
    beforeAll(() => {
      loginInfoStorage.set(loginInfo);

      return () => loginInfoStorage.remove();
    });

    it("should render the login page with a password field", () => {
      const { container } = render(Login, {});

      expect(container.firstChild).toMatchSnapshot();
    });

    it("should show an error message if the user enters the wrong password", async () => {
      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      expect(getErrorElement()).toBeNull();

      await fireEvent.input(textInput, { target: { value: "foo bar" } });
      await fireEvent.submit(form, { currentTarget: form });

      const errorElement = await vi.waitUntil(getErrorElement);
      const selectedText = textInput.value.substring(
        Number(textInput.selectionStart),
        Number(textInput.selectionEnd)
      );

      expect(initSpy).not.toHaveBeenCalled();
      expect(errorElement?.textContent).toMatch(/password/i);
      expect(textInput).toHaveFocus();
      expect(selectedText).toBe(textInput.value);
    });

    /**
     * This is not a possible situation, in theory, but
     * the workflow is able to deal with it.
     */
    it("should trigger the Reset flow if the user inputs the correct password with no prior wallet created", async () => {
      settingsStore.update(setKey("userId", ""));

      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: pwd } });
      await fireEvent.submit(form, { currentTarget: form });

      expect(getErrorElement()).toBeNull();

      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getErrorElement()).toBeNull();
      expect(getWalletSpy).toHaveBeenCalledTimes(1);
      expect(getWalletSpy).toHaveBeenCalledWith(seed);
      expect(initSpy).not.toHaveBeenCalled();
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/setup/restore");
    });

    /**
     * This is not a possible situation, in theory, but
     * the workflow is able to deal with it.
     */
    it("should trigger the Reset flow if the user inputs the correct password for a mnemonic different from the last one used", async () => {
      settingsStore.update(setKey("userId", "some-fake-id"));

      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: pwd } });
      await fireEvent.submit(form, { currentTarget: form });

      expect(getErrorElement()).toBeNull();

      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getWalletSpy).toHaveBeenCalledTimes(1);
      expect(getWalletSpy).toHaveBeenCalledWith(seed);
      expect(initSpy).not.toHaveBeenCalled();
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/setup/restore");
    });

    it("should redirect to the Dashboard if the entered password is for the last used mnemonic", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, { target: { value: pwd } });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getWalletSpy).toHaveBeenCalledTimes(1);
      expect(getWalletSpy).toHaveBeenCalledWith(seed);
      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(Wallet));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });

    it("should trim the entered password before validating it", async () => {
      settingsStore.update(setKey("userId", userId));

      const { container } = render(Login, {});
      const form = getAsHTMLElement(container, "form");
      const textInput = getTextInput(container);

      await fireEvent.input(textInput, {
        target: { value: `  \t${pwd} \t  ` },
      });
      await fireEvent.submit(form, { currentTarget: form });
      await vi.waitUntil(() => gotoSpy.mock.calls.length > 0);

      expect(getWalletSpy).toHaveBeenCalledTimes(1);
      expect(getWalletSpy).toHaveBeenCalledWith(seed);
      expect(get(settingsStore).userId).toBe(userId);
      expect(initSpy).toHaveBeenCalledTimes(1);
      expect(initSpy).toHaveBeenCalledWith(expect.any(Wallet));
      expect(gotoSpy).toHaveBeenCalledTimes(1);
      expect(gotoSpy).toHaveBeenCalledWith("/dashboard");
    });
  });
});
