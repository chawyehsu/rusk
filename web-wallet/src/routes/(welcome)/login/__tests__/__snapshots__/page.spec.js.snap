// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login > Mnemonic phrase workflow > should render the login page with a mnemonic phrase field 1`] = `
<div>
  <section
    class="landing-content s-diEOsynXV9Ac"
  >
    <h2
      class="h1 s-diEOsynXV9Ac"
    >
      Unlocking the Future: 
      <br
        class="s-diEOsynXV9Ac"
      />
      
    Your 
      <mark
        class="s-diEOsynXV9Ac"
      >
        DUSK
      </mark>
       Native Wallet
    </h2>
     
    <div
      class="landing-content__options s-diEOsynXV9Ac"
    >
      <form
        class="s-diEOsynXV9Ac"
      >
        <div
          class="dusk-field-button-group"
        >
          <input
            autocomplete="current-password"
            class="dusk-textbox dusk-textbox-password"
            name="secret"
            placeholder="Mnemonic phrase"
            required=""
            tabindex="0"
            type="password"
          />
          
          <!--&lt;Textbox&gt;-->
           
          <button
            class="dusk-button dusk-button--type--submit dusk-button--variant--secondary dusk-button--size--normal"
            type="submit"
          >
            <span
              class="dusk-button__text"
            >
              Unlock
            </span>
          </button>
          <!--&lt;Button&gt;-->
        </div>
        <!--&lt;FieldButtonGroup&gt;-->
         
      </form>
       
      <div
        class="landing-content__options-group s-diEOsynXV9Ac"
      >
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--normal dusk-icon-button-labeled"
          href="/some-base-path/setup/create"
        >
          <svg
            class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M5,3C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V16.72C21.59,16.37 22,15.74 22,15V9C22,8.26 21.59,7.63 21,7.28V5A2,2 0 0,0 19,3H5M5,5H19V7H13A2,2 0 0,0 11,9V15A2,2 0 0,0 13,17H19V19H5V5M13,9H20V15H13V9M16,10.5A1.5,1.5 0 0,0 14.5,12A1.5,1.5 0 0,0 16,13.5A1.5,1.5 0 0,0 17.5,12A1.5,1.5 0 0,0 16,10.5Z"
            />
          </svg>
          
          <!--&lt;Icon&gt;-->
           
          <span
            class="dusk-anchor-button__text"
          >
            New
          </span>
          
          
        </a>
        <!--&lt;Anchor&gt;-->
        <!--&lt;AnchorButton&gt;-->
        <!--&lt;AppAnchorButton&gt;-->
         
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--normal dusk-icon-button-labeled"
          href="/some-base-path/setup/restore"
        >
          <svg
            class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M13,3A9,9 0 0,0 4,12H1L4.89,15.89L4.96,16.03L9,12H6A7,7 0 0,1 13,5A7,7 0 0,1 20,12A7,7 0 0,1 13,19C11.07,19 9.32,18.21 8.06,16.94L6.64,18.36C8.27,20 10.5,21 13,21A9,9 0 0,0 22,12A9,9 0 0,0 13,3Z"
            />
          </svg>
          
          <!--&lt;Icon&gt;-->
           
          <span
            class="dusk-anchor-button__text"
          >
            Restore
          </span>
          
          
        </a>
        <!--&lt;Anchor&gt;-->
        <!--&lt;AnchorButton&gt;-->
        <!--&lt;AppAnchorButton&gt;-->
      </div>
    </div>
  </section>
   
  <footer
    class="landing-footer s-diEOsynXV9Ac"
  >
    <span
      class="s-diEOsynXV9Ac"
    >
      Web Wallet v
      0.1.5
       (
      hash1234 2024-01-12
      )
    </span>
  </footer>
  <!--&lt;+page&gt;-->
</div>
`;

exports[`Login > Password workflow > should render the login page with a password field 1`] = `
<div>
  <section
    class="landing-content s-diEOsynXV9Ac"
  >
    <h2
      class="h1 s-diEOsynXV9Ac"
    >
      Unlocking the Future: 
      <br
        class="s-diEOsynXV9Ac"
      />
      
    Your 
      <mark
        class="s-diEOsynXV9Ac"
      >
        DUSK
      </mark>
       Native Wallet
    </h2>
     
    <div
      class="landing-content__options s-diEOsynXV9Ac"
    >
      <form
        class="s-diEOsynXV9Ac"
      >
        <div
          class="dusk-field-button-group"
        >
          <input
            autocomplete="current-password"
            class="dusk-textbox dusk-textbox-password"
            name="secret"
            placeholder="Password"
            required=""
            tabindex="0"
            type="password"
          />
          
          <!--&lt;Textbox&gt;-->
           
          <button
            class="dusk-button dusk-button--type--submit dusk-button--variant--secondary dusk-button--size--normal"
            type="submit"
          >
            <span
              class="dusk-button__text"
            >
              Unlock
            </span>
          </button>
          <!--&lt;Button&gt;-->
        </div>
        <!--&lt;FieldButtonGroup&gt;-->
         
      </form>
       
      <div
        class="landing-content__options-group s-diEOsynXV9Ac"
      >
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--primary dusk-anchor-button--size--normal dusk-icon-button-labeled"
          href="/some-base-path/setup/create"
        >
          <svg
            class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M5,3C3.89,3 3,3.9 3,5V19A2,2 0 0,0 5,21H19A2,2 0 0,0 21,19V16.72C21.59,16.37 22,15.74 22,15V9C22,8.26 21.59,7.63 21,7.28V5A2,2 0 0,0 19,3H5M5,5H19V7H13A2,2 0 0,0 11,9V15A2,2 0 0,0 13,17H19V19H5V5M13,9H20V15H13V9M16,10.5A1.5,1.5 0 0,0 14.5,12A1.5,1.5 0 0,0 16,13.5A1.5,1.5 0 0,0 17.5,12A1.5,1.5 0 0,0 16,10.5Z"
            />
          </svg>
          
          <!--&lt;Icon&gt;-->
           
          <span
            class="dusk-anchor-button__text"
          >
            New
          </span>
          
          
        </a>
        <!--&lt;Anchor&gt;-->
        <!--&lt;AnchorButton&gt;-->
        <!--&lt;AppAnchorButton&gt;-->
         
        <a
          aria-disabled="false"
          class="dusk-anchor dusk-anchor-button dusk-anchor-button--variant--tertiary dusk-anchor-button--size--normal dusk-icon-button-labeled"
          href="/some-base-path/setup/restore"
        >
          <svg
            class="dusk-icon dusk-icon--size--normal dusk-anchor-button__icon"
            role="graphics-symbol"
            viewBox="0 0 24 24"
          >
            <path
              d="M13,3A9,9 0 0,0 4,12H1L4.89,15.89L4.96,16.03L9,12H6A7,7 0 0,1 13,5A7,7 0 0,1 20,12A7,7 0 0,1 13,19C11.07,19 9.32,18.21 8.06,16.94L6.64,18.36C8.27,20 10.5,21 13,21A9,9 0 0,0 22,12A9,9 0 0,0 13,3Z"
            />
          </svg>
          
          <!--&lt;Icon&gt;-->
           
          <span
            class="dusk-anchor-button__text"
          >
            Restore
          </span>
          
          
        </a>
        <!--&lt;Anchor&gt;-->
        <!--&lt;AnchorButton&gt;-->
        <!--&lt;AppAnchorButton&gt;-->
      </div>
    </div>
  </section>
   
  <footer
    class="landing-footer s-diEOsynXV9Ac"
  >
    <span
      class="s-diEOsynXV9Ac"
    >
      Web Wallet v
      0.1.5
       (
      hash1234 2024-01-12
      )
    </span>
  </footer>
  <!--&lt;+page&gt;-->
</div>
`;
